source/atomic_thread_fence.rst

```diff
@@ -0,0 +1,96 @@
+          int const myepoch = epoch.load(memory_order_relaxed);
+          int const result = arrived.fetch_add(1, memory_order_acq_rel) + 1;
+          if(result == expected) {
+              expected = nexpected.load(memory_order_relaxed);
+              arrived.store(0, memory_order_relaxed);
+              epoch.store(myepoch + 1, memory_order_release);
+          }
+          else
+              while(epoch.load(memory_order_acquire) == myepoch)
+                  ;
+      }
+  private:
+      int expected;
+      atomic<int> arrived, nexpected, epoch;
+  };
+
+The release operation on the epoch atomic is likely to require the compiler to
+insert a fence that has an effect that goes beyond the intended constraint,
+which is to order only the operations on the barrier object. Since the barrier
+object is likely to be smaller than a cache line and the library's
+implementation can control its alignment using ``alignas``, then it would be
+possible to compile this program without a fence in this location on
+architectures that are cache-line coherent. To concisely express the bound on
+the set of memory operations whose order is constrained, we propose to overload
+``std::atomic_thread_fence`` with a variant that takes a reference to the object
+containing sub-objects to be ordered by the fence:
+
+.. code-block:: c++
+
+     template<class... T>
+     void atomic_thread_fence(memory_order, T... &&objects) noexcept;
+
+For this overload, operations on objects that are not sub-objects of the
+object(s) in the variadic template argument(s) are *un-sequenced* with the
+fence. The compiler would likely apply restrictions on alignment, and may
+generate a dynamic test leading to a fence for under-aligned objects.
+
+In all cases, a trivially conforming implementation may implement the new
+overload in terms of the existing ``std::atomic_thread_fence`` using the same
+memory order:
+
+.. code-block:: c++
+
+     template<class... T>
+     void atomic_thread_fence(memory_order, T... &&) noexcept {
+       atomic_thread_fence(memory_order);
+     }
+
+The above barrier example's inner-code would use the new overload as follows:
+
+.. code-block:: c++
+
+          if(result == expected) {
+              expected = nexpected.load(memory_order_relaxed);
+              arrived.store(0, memory_order_relaxed);
+	      atomic_thread_fence(memory_order_release, *this);
+              epoch.store(myepoch + 1, memory_order_relaxed);
+          }
+
+This enables hardware-specific optimizations which cannot be expressed in C++
+today. If the synchronized object(s) are know to reside in memory that's not
+visible to other threads of execution, then a weaker type of fence than the
+hardware's global fence can be used.
+
+We conjecture that this mechanism could also be used to identify transaction
+boundaries in hardware implementations which support transactional memory, and
+implement subsequent atomic operations as hardware transactions instead.
```

Comment:
We should discuss how this paper's proposal is be similar to consume in that it expects the compiler to use dependency edges, which an SSA compiler doesn't usually do. It's different from consume because it's very bounded in dependencies, and doesn't expect them to propagate down to the machine-level code generation! It looks like a barrier opcode in IR, but has extra metadata detailing memory locations.